packages/component-base/src/slot-observer.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright (c) 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * A helper for observing slot changes.
+ */
+export class SlotObserver {
+  constructor(slot, callback) {
+    /** @type HTMLSlotElement */
+    this.slot = slot;
+
+    /** @type Function */
+    this.callback = callback;
+
+    /** @type {Node[]} */
+    this._storedNodes = [];
+
+    this._processNodes();
+
+    slot.addEventListener('slotchange', () => {
+      this._processNodes();
+    });
+  }
+
+  /**
+   * Run the observer callback synchronously.
+   */
+  flush() {
+    this._processNodes();
+  }
+
+  /** @private */
+  _processNodes() {
+    const currentNodes = this.slot.assignedNodes({ flatten: true });
+
+    let addedNodes = [];
+    let removedNodes = [];
+
+    if (currentNodes.length) {
+      addedNodes = currentNodes.filter((node) => !this._storedNodes.includes(node));
+    }
+
+    if (this._storedNodes.length) {
+      removedNodes = this._storedNodes.filter((node) => !currentNodes.includes(node));
+    }
```

Comment:
This logic is problematic as it doesn't cover cases where all the nodes are still there, but their order has changed. So, for example, the following `vaadin-accordion` test fails with `SlotObserver`:

https://github.com/vaadin/web-components/blob/73db22a5e8993e3ce48d1e6540d30eff9cb5c257/packages/accordion/test/accordion.test.js#L78-L84

Reply:
This should be now addressed by adding `movedNodes`.